File: kotlin/isbn-verifier/src/main/kotlin/IsbnVerifier.kt
class IsbnVerifier {
    fun isValid(isbn: String) = isbn.normalized().let { it.hasValidFormat() && it.hasValidChecksum() }

    private fun String.normalized() = replace("-", "")
    private fun String.hasValidFormat() = matches("""^\d{9}(\d|X)$""".toRegex())
    private fun String.hasValidChecksum() = checksum().rem(11) == 0
    private fun String.checksum() = mapIndexed { i, c -> c.toDigit().times(10 - i) }.sum()

    private fun Char.toDigit() = if (this == 'X') 10 else Character.getNumericValue(this)
}

File: nim/isbn-verifier/isbn_verifier.nim
import re, strutils

let validIsbn = re"^(\d{9}(\d|X))$"

proc removeDashes(isbn: string): string =
  isbn.replace("-", "")

proc digit(c: char): int =
  if c == 'X': 10 else: parseInt($c)

proc checksum(isbn: string): int =
  for i in countdown(isbn.len, 1):
    result += isbn[i - 1].digit * i

proc isValid*(isbn: string): bool =
  if isbn.removeDashes =~ validIsbn:
    let isbn = matches[0]
    isbn.checksum mod 11 == 0
  else:
    false
